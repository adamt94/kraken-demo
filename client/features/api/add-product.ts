export const fakeAddProductRequest = async (productId: string) => {
  return new Promise((resolve) => {
    setTimeout(() => {
      console.log(`Product ${productId} added!`)
      resolve(true)
    }, 500)
  })
}
