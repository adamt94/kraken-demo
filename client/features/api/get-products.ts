import { gql } from "@apollo/client";

export const GET_PRODUCTS = gql`
  query {
    allProducts {
      id
      name
      price
      height
      width
      length
      model_code
      colour
      img_url
      power
      quantity
      description
      brand
    }
  }
`;
