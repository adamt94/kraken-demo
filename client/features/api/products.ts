import { gql } from "@apollo/client";

export type ProductType = {
  id: string;
  name: string;
  price: number;
  height: number;
  width: number;
  length: number;
  model_code: string;
  colour: string;
  img_url: string;
  power: string;
  quantity: number;
  description: string;
  brand: string;
};

export const GET_PRODUCTS = gql`
  query {
    allProducts {
      id
      name
      price
      height
      width
      length
      model_code
      colour
      img_url
      power
      quantity
      description
      brand
    }
  }
`;

export const GET_PRODUCT = gql`
  query GetProduct($id: ID!) {
    Product(id: $id) {
      id
      name
      price
      height
      width
      length
      model_code
      colour
      img_url
      power
      quantity
      description
      brand
    }
  }
`;

