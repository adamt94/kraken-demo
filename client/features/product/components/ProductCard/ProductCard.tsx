"use client"

import Image from "next/image"
import { QuantityInput } from "../QuantityInput/QuantityInput";
import { Button } from "@/components/Button/Button";
import { fakeAddProductRequest } from "@/features/api/add-product";

type ProductCardProps = {
  image: string;
  title: string;
  subtitle: string;
  price: number;
}


export const ProductCard = ({ image, title, subtitle, price }: ProductCardProps) => {

  const handleClick = async () => {
    try {
      await fakeAddProductRequest("");
    } catch (error) {
      console.error("Error adding product:", error);
    }
  };

  return (
    <div className="px-6 md:grid md:grid-cols-2 lg:grid-cols-5 gap-8 lg:gap-12">
      <div className="rounded-lg mb-6 md:mb-0 lg:col-span-2">
        <div className="flex justify-center">
          <Image
            src={image}
            alt={title}
            width={500}
            height={500}
            className="object-contain rounded-3xl"
            priority
          />
        </div>
      </div>
      <div className="lg:col-span-3 flex flex-col">
        <div className="space-y-2 mb-6">
          <h1 className="text-4xl md:text-5xl leading-13">{title}</h1>
          <p className="text-onBackground text-xl">{subtitle}</p>
        </div>
        <div className="flex flex-row md:flex-row justify-between items-end gap-4 mb-6">
          <div className="text-3xl md:text-4xl font-bold">Â£{price}</div>
          <div className="flex flex-col items-center gap-2">
            <span className="text-onBackground text-sm">Qty</span>
            <QuantityInput />
          </div>
        </div>
        <Button
          className="py-6 text-lg"
          onClick={handleClick}
        >
          Add to cart
        </Button>

      </div>

    </div>
  );
}
