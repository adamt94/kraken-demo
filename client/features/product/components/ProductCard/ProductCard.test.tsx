import React from 'react';
import { render, screen, waitFor, act } from '@testing-library/react';
import { ProductCard } from "./ProductCard";
import { fakeAddProductRequest } from '@/features/api/add-product';

jest.mock('../../../api/add-product', () => ({
  fakeAddProductRequest: jest.fn(),
}))

describe('ProductCard', () => {
  it('renders price, quantity, title and subtitle', () => {
    render(
      <ProductCard
        image="https://example.com/image.jpg"
        title="Test Product"
        subtitle="This is a test product"
        price={19.99}
      />
    );
    screen.getByText('Test Product');
    screen.getByText('This is a test product');
    screen.getByText('Qty');
    screen.getByText('Â£19.99');
  });

  it('handles button click', async () => {

    const { getByText } = render(
      <ProductCard
        image="https://example.com/image.jpg"
        title="Test Product"
        subtitle="This is a test product"
        price={19.99}
      />
    );

    const button = getByText('Add to cart');

    act(() => {
      button.click();
    });

    await waitFor(() => {
      expect(fakeAddProductRequest).toHaveBeenCalled();
    });
  });
});
