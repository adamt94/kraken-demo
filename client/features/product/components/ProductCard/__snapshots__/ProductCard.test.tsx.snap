// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard renders correctly 1`] = `
<div>
  <div
    class="md:grid md:grid-cols-2 lg:grid-cols-5 gap-8 lg:gap-12"
  >
    <div
      class="rounded-lg p-4 mb-6 md:mb-0 lg:col-span-2"
    >
      <div
        class="flex justify-center"
      >
        <img
          alt="Test Product"
          class="object-contain rounded-3xl"
          data-nimg="1"
          decoding="async"
          height="500"
          src="/_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=1080&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=1080&q=75 2x"
          style="color: transparent;"
          width="500"
        />
      </div>
    </div>
    <div
      class="lg:col-span-3 flex flex-col"
    >
      <div
        class="space-y-2 mb-6"
      >
        <h1
          class="text-4xl md:text-5xl leading-13"
        >
          Test Product
        </h1>
        <p
          class="text-onBackground text-xl"
        >
          This is a test product
        </p>
      </div>
      <div
        class="flex flex-row md:flex-row justify-between items-end gap-4 mb-6"
      >
        <div
          class="text-3xl md:text-4xl font-bold"
        >
          Â£
          19.99
        </div>
        <div
          class="flex flex-col items-center gap-2"
        >
          <span
            class="text-onBackground text-sm"
          >
            Qty
          </span>
          <div
            class="flex items-center"
          >
            <button
              class="inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-onPrimary hover:bg-primary/90 h-10 w-10 rounded-xl text-xl"
              disabled=""
            >
              -
            </button>
            <span
              class="mx-3 text-2xl w-4 text-center"
            >
              1
            </span>
            <button
              class="inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-onPrimary hover:bg-primary/90 h-10 w-10 rounded-xl text-xl"
            >
              +
            </button>
          </div>
        </div>
      </div>
      <button
        class="inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-onPrimary hover:bg-primary/90 h-10 px-4 py-6 text-lg"
      >
        Add to cart
      </button>
    </div>
  </div>
</div>
`;
